specs/heading_attrs.txt

```diff
@@ -0,0 +1,554 @@
+````````````````````````````````
+
+There are no deduplication of classes.
+
+```````````````````````````````` example
+# H1 {.a .a .a}
+.
+<h1 class="a a a">H1</h1>
+````````````````````````````````
+
+## Combined
+
+When both an ID and classes are specified, ID attribute is always emitted first.
+
+```````````````````````````````` example
+# H1 {.myclass #myid}
+## H2 {.z #m .a}
+.
+<h1 id="myid" class="myclass">H1</h1>
+<h2 id="m" class="z a">H2</h2>
+````````````````````````````````
+
+# Unknown attributes
+
+Unknown attributes are simply ignored.
+Note that it cause neither an error nor a fallback to non-extended markdown,
+since this other attributes can be supported by other implementations or this
+crate in future.
+
+```````````````````````````````` example
+# H1 {foo}
+## H2 {#myid unknown this#is.ignored .myclass}
+.
+<h1>H1</h1>
+<h2 id="myid" class="myclass">H2</h1>
+````````````````````````````````
+
+# Forbidden characters
+
+Some characters cannot appear in attribute blocks.
+Forbidden characters are:
+
+* left curly brace (`{`),
+* right curly brace (`}`),
+* less-than operator (`<`),
+* greater-than operator (`>`),
+* backslash (`\`), and
+* newline.
+
+Left curly brace (note that `{unknown}` and `{.bar}` are treated as attribute blocks):
+
+```````````````````````````````` example
+# H1 {.foo{unknown}
+## H2 {.foo{.bar}
+.
+<h1>H1 {.foo</h1>
+<h2 class="bar">H2 {.foo</h2>
+````````````````````````````````
+
+Right curly brace:
+
+```````````````````````````````` example
+# H1 {.foo}bar}
+.
+<h1>H1 {.foo}bar}</h1>
+````````````````````````````````
+
+less-than operator and greater-than operator:
+
+```````````````````````````````` example
+# H1 {<i>foo</i>}
+.
+<h1>H1 {<i>foo</i>}</h1>
+````````````````````````````````
+
+Backslash:
+
+```````````````````````````````` example
+# H1 {.foo\}
+.
+<h1>H1 {.foo}</h1>
+````````````````````````````````
+
+Newline:
+
+```````````````````````````````` example
+H1 {.foo
+.bar}
+==
+.
+<h1>H1 {.foo
+.bar}</h1>
+````````````````````````````````
+
+# Cancelling parsing of attribute blocks
+
+If you want to avoid the last `{...}` from being treated as an attribute block,
+you can do it in some ways.
+
+Put a dummy empty attribute block `{}` at the end of line:
+
+```````````````````````````````` example
+H1 {} {}
+=====
+
+## H2 {} {}
+.
+<h1>H1 {}</h1>
+<h2>H2 {}</h2>
+````````````````````````````````
+
+Or, put closing hashes for ATX heading:
+
+```````````````````````````````` example
+## H2 {} ##
+.
+<h2>H2 {}</h2>
+````````````````````````````````
+
+Or, include forbidden characters (a backslash and newline in this example)
+between the braces.
+
+```````````````````````````````` example
+# H1 {\}
+## this is also ok \{\}
+
+newline can be used for setext heading {
+}
+--
+.
+<h1>H1 {}</h1>
+<h2>this is also ok {}</h2>
+<h2>newline can be used for setext heading {
+}</h2>
+````````````````````````````````
+
+Note that escaping only opening braces does not prevent the braces from
```

Comment:
I know this would be a pain to parse properly, but this is mildly surprising. Do other parsers have the same behavior?

Reply:
[babelmark3 | Compare Markdown Implementations](https://babelmark.github.io/?text=H1+%7B+%23id1+%7D%0A%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%0A%0AH1+%7B+%23id2%0A%7D%0A%3D%3D%3D%3D%3D%3D%3D%3D%3D)

No parsers seems to be able to recognize multi-line attribute blocks.
I don't know how other implementations not supported by babelmark3 would handle such cases, but I think it would be safe to assume no users write attribute blocks in multiline form (as processors cannot handle such attributes).

Reply:
Woops, I meant escaping opening braces!

Reply:
[babelmark3 | Compare Markdown Implementations](https://babelmark.github.io/?text=%23+H1+%7B+%23id1+%7D%0A%23%23+H2+%5C%7B+%23id2+%7D)

* Opening braces can be escaped: php-markdown-extra 1.9.1, pandoc 2.16.2
* Escape doesn't work (same as current impl.): markdig (advanced) 0.26.0.0, maruku 0.7.3.beta1

There seems no common behavior...

It would be natural if `\{ ... }` is not parsed as an attribute block.
And it'll not so hard to support such escapes, since `{` itself is prohibited in attribute blocks and we don't need to care multiple escapes in various contexts.

I have no strong preference for how this should be parsed. Both of "attribute blocks are trimmed first, before parsing" and "escaped character doesn't behave specially" sounds natural to me.

Reply:
Ah, ok, I remembered why I avoided supporting escaping.

As I wrote at <https://github.com/raphlinus/pulldown-cmark/pull/522#issuecomment-971850546>, it is hard to decide how `# h1 {.foo\{.bar}` should be converted.

* Users' expectation with escaping support would be: content=`h1`, class=`foo{.bar`.
* However, with escaping support, this would be converted to: content=`h1 {.foo{.bar}`, class=None.
    + This is because `{` cannot appear in attribute blocks.
* Current implementation: content=`h1 {.foo\`, class=`bar`.

I don't like this ambiguity or unexpectedness when escaping opening braces is allowed, so I decided to trim the block before processing any escapes.
But I'm not confident this is better, so I'll change implementation if you have some preference.